From f1a85710ab909fb1d0c5559b36a1d9d1fe0d8eb1 Mon Sep 17 00:00:00 2001
From: Lukas Czerner <lczerner@redhat.com>
Date: Thu, 23 Oct 2014 09:53:37 -0400
Subject: [fs] ext4: isolate ext4_extents.h file

Message-id: <1414058125-4183-64-git-send-email-lczerner@redhat.com>
Patchwork-id: 98170
O-Subject: [RHEL 7.1 PATCH v2 063/171] ext4: isolate ext4_extents.h file
Bugzilla: 1150139
RH-Acked-by: Eric Sandeen <sandeen@redhat.com>

From: Zheng Liu <wenqing.lz@taobao.com>

BZ 1150139
https://bugzilla.redhat.com/show_bug.cgi?id=1150139

BREW 8142710
https://brewweb.devel.redhat.com/taskinfo?taskID=8142710

Upstream commit d7b2a00c2e2eedf460ce2a15237f28de40412d86

 After applied the commit (4a092d73), we have reduced the number of
 source files that need to #include ext4_extents.h.  But we can do
 better.

 This commit defines ext4_zeroout_es() in extents.c and move
 EXT_MAX_BLOCKS into ext4.h in order not to include ext4_extents.h in
 indirect.c and ioctl.c.  Meanwhile we just need to include this file in
 extent_status.c when ES_AGGRESSIVE_TEST is defined.  Otherwise, this
 commit removes a duplicated declaration in trace/events/ext4.h.

 After applied this patch, we just need to include ext4_extents.h file
 in {super,migrate,move_extents,extents}.c, and it is easy for us to
 define a new extent disk layout.

 Signed-off-by: Zheng Liu <wenqing.lz@taobao.com>
 Signed-off-by: "Theodore Ts'o" <tytso@mit.edu>
 Signed-off-by: Lukas Czerner <lczerner@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/ext4/ext4.h b/fs/ext4/ext4.h
index c16f6cc..03edf84 100644
--- a/fs/ext4/ext4.h
+++ b/fs/ext4/ext4.h
@@ -2692,6 +2692,12 @@ extern int ext4_check_blockref(const char *, unsigned int,
 struct ext4_ext_path;
 struct ext4_extent;
 
+/*
+ * Maximum number of logical blocks in a file; ext4_extent's ee_block is
+ * __le32.
+ */
+#define EXT_MAX_BLOCKS 0xffffffff
+
 extern int ext4_ext_tree_init(handle_t *handle, struct inode *);
 extern int ext4_ext_writepage_trans_blocks(struct inode *, int);
 extern int ext4_ext_index_trans_blocks(struct inode *inode, int extents);
diff --git a/fs/ext4/ext4_extents.h b/fs/ext4/ext4_extents.h
index 51bc821..5074fe2 100644
--- a/fs/ext4/ext4_extents.h
+++ b/fs/ext4/ext4_extents.h
@@ -134,12 +134,6 @@ struct ext4_ext_path {
  */
 
 /*
- * Maximum number of logical blocks in a file; ext4_extent's ee_block is
- * __le32.
- */
-#define EXT_MAX_BLOCKS 0xffffffff
-
-/*
  * EXT_INIT_MAX_LEN is the maximum number of blocks we can have in an
  * initialized extent. This is 2^15 and not (2^16 - 1), since we use the
  * MSB of ee_len field in the extent datastructure to signify if this
diff --git a/fs/ext4/extents.c b/fs/ext4/extents.c
index 8e7f1e7..74e6897 100644
--- a/fs/ext4/extents.c
+++ b/fs/ext4/extents.c
@@ -3048,6 +3048,23 @@ void ext4_ext_release(struct super_block *sb)
 #endif
 }
 
+static int ext4_zeroout_es(struct inode *inode, struct ext4_extent *ex)
+{
+ ext4_lblk_t  ee_block;
+ ext4_fsblk_t ee_pblock;
+ unsigned int ee_len;
+
+ ee_block  = le32_to_cpu(ex->ee_block);
+ ee_len    = ext4_ext_get_actual_len(ex);
+ ee_pblock = ext4_ext_pblock(ex);
+
+ if (ee_len == 0)
+  return 0;
+
+ return ext4_es_insert_extent(inode, ee_block, ee_len, ee_pblock,
+         EXTENT_STATUS_WRITTEN);
+}
+
 /* FIXME!! we need to try to merge to left or right after zero-out  */
 static int ext4_ext_zeroout(struct inode *inode, struct ext4_extent *ex)
 {
@@ -3200,7 +3217,7 @@ static int ext4_split_extent_at(handle_t *handle,
    goto fix_extent_len;
 
   /* update extent status tree */
-  err = ext4_es_zeroout(inode, &zero_ex);
+  err = ext4_zeroout_es(inode, &zero_ex);
 
   goto out;
  } else if (err)
@@ -3551,7 +3568,7 @@ static int ext4_ext_convert_to_initialized(handle_t *handle,
 out:
  /* If we have gotten a failure, don't zero out status tree */
  if (!err)
-  err = ext4_es_zeroout(inode, &zero_ex);
+  err = ext4_zeroout_es(inode, &zero_ex);
  return err ? err : allocated;
 }
 
diff --git a/fs/ext4/extents_status.c b/fs/ext4/extents_status.c
index 0e88a36..2d1bdbe 100644
--- a/fs/ext4/extents_status.c
+++ b/fs/ext4/extents_status.c
@@ -13,7 +13,6 @@
 #include <linux/list_sort.h>
 #include "ext4.h"
 #include "extents_status.h"
-#include "ext4_extents.h"
 
 #include <trace/events/ext4.h>
 
@@ -409,6 +408,8 @@ ext4_es_try_to_merge_right(struct inode *inode, struct extent_status *es)
 }
 
 #ifdef ES_AGGRESSIVE_TEST
+#include "ext4_extents.h" /* Needed when ES_AGGRESSIVE_TEST is defined */
+
 static void ext4_es_insert_extent_ext_check(struct inode *inode,
          struct extent_status *es)
 {
@@ -903,23 +904,6 @@ int ext4_es_remove_extent(struct inode *inode, ext4_lblk_t lblk,
  return err;
 }
 
-int ext4_es_zeroout(struct inode *inode, struct ext4_extent *ex)
-{
- ext4_lblk_t  ee_block;
- ext4_fsblk_t ee_pblock;
- unsigned int ee_len;
-
- ee_block  = le32_to_cpu(ex->ee_block);
- ee_len    = ext4_ext_get_actual_len(ex);
- ee_pblock = ext4_ext_pblock(ex);
-
- if (ee_len == 0)
-  return 0;
-
- return ext4_es_insert_extent(inode, ee_block, ee_len, ee_pblock,
-         EXTENT_STATUS_WRITTEN);
-}
-
 static int ext4_inode_touch_time_cmp(void *priv, struct list_head *a,
          struct list_head *b)
 {
diff --git a/fs/ext4/extents_status.h b/fs/ext4/extents_status.h
index 3e83aef..167f4ab 100644
--- a/fs/ext4/extents_status.h
+++ b/fs/ext4/extents_status.h
@@ -81,7 +81,6 @@ extern void ext4_es_find_delayed_extent_range(struct inode *inode,
      struct extent_status *es);
 extern int ext4_es_lookup_extent(struct inode *inode, ext4_lblk_t lblk,
      struct extent_status *es);
-extern int ext4_es_zeroout(struct inode *inode, struct ext4_extent *ex);
 
 static inline int ext4_es_is_written(struct extent_status *es)
 {
diff --git a/fs/ext4/indirect.c b/fs/ext4/indirect.c
index 87b30cd..594009f 100644
--- a/fs/ext4/indirect.c
+++ b/fs/ext4/indirect.c
@@ -23,7 +23,6 @@
 #include <linux/aio.h>
 #include "ext4_jbd2.h"
 #include "truncate.h"
-#include "ext4_extents.h" /* Needed for EXT_MAX_BLOCKS */
 
 #include <trace/events/ext4.h>
 
diff --git a/fs/ext4/ioctl.c b/fs/ext4/ioctl.c
index 527698d..60589b6 100644
--- a/fs/ext4/ioctl.c
+++ b/fs/ext4/ioctl.c
@@ -17,7 +17,6 @@
 #include <asm/uaccess.h>
 #include "ext4_jbd2.h"
 #include "ext4.h"
-#include "ext4_extents.h"
 
 #define MAX_32_NUM ((((unsigned long long) 1) << 32) - 1)
 
diff --git a/include/trace/events/ext4.h b/include/trace/events/ext4.h
index d4e483e..deb12ac 100644
--- a/include/trace/events/ext4.h
+++ b/include/trace/events/ext4.h
@@ -14,7 +14,6 @@ struct ext4_prealloc_space;
 struct ext4_inode_info;
 struct mpage_da_data;
 struct ext4_map_blocks;
-struct ext4_extent;
 struct extent_status;
 
 #define EXT4_I(inode) (container_of(inode, struct ext4_inode_info, vfs_inode))
-- 
1.7.1